src/compiler/binder.ts

```diff
@@ -2803,18 +2738,20 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
                 }
                 // falls through
             case SyntaxKind.ThisKeyword:
+                // TODO: Why use `isExpression` here? both Identifier and ThisKeyword are expressions.

```

Comment:
IIRC `this` in a type position wasn't an expression, but was still parsed as a `ThisKeyword` for a period of time.